packages/hooks_riverpod/lib/src/user_provider_listener.dart

```diff
@@ -0,0 +1,148 @@
+import 'package:flutter/foundation.dart';
+import 'package:flutter/widgets.dart';
+import 'package:flutter_hooks/flutter_hooks.dart';
+import 'package:flutter_riverpod/flutter_riverpod.dart';
+
+import 'internals.dart';
+
+/// A hook that listens to a provider and and triggers a callback whenevr the value of provider is changed.
+///
+/// As a usage example, consider:
+///
+/// ```dart
+/// class Counter extends StateNotifier<int> {
+///   Counter(): super(0);
+///    void increment() => state++;
+/// }
+///
+/// final counterProvider = StateNotifierProvider((ref) => Counter());
+/// ```
+///
+/// We can then use [useProviderListener] to listen to `helloWorldProvider` like so:
+///
+/// ```dart
+/// class Example extends HookWidget {
+///   @override
+///   Widget build(BuildContext context) {
+///     final counter = userProvider(counter.state);
+///     useProviderListener(counterProvider, (context, value) {
+///       if (value % 2 == 0) print('even');
+///       else print('odd');
+///     });
+///    return Column(
+///       children: [
+///         Text('$count'),
+///         RaisedButton(
+///           onPressed: () => context.read(counterProvider).increment(),
+///          child: const Text('increment'),
+///         )
+///       ],
+///     );
+///   }
+/// }
+/// ```
+///
+/// See also:
+///
+/// - [Provider]/`select`, for filtering unwanted rebuilds.
+T useProviderListener<T>(
+    ProviderListenable<T> provider, OnProviderChange<T> onChange) {
+  final container = ProviderScope.containerOf(useContext());
+  return use(_ProviderListenerHook<T>(container, provider, onChange));
+}
+
+typedef OnProviderChange<T> = void Function(BuildContext context, T value);
+
+class _ProviderListenerHook<T> extends Hook<T> {
+  const _ProviderListenerHook(
+      this._container, this._providerListenable, this.onChange);
+
+  final ProviderContainer _container;
+  final ProviderListenable<T> _providerListenable;
+  final OnProviderChange<T> onChange;
+
+  @override
+  _ProviderListenerHookState<T> createState() => _ProviderListenerHookState();
+
+  @override
+  void debugFillProperties(DiagnosticPropertiesBuilder properties) {
+    super.debugFillProperties(properties);
+    properties.add(
+      DiagnosticsProperty<OnProviderChange<T>>('onChange', onChange),
+    );
+    properties.add(DiagnosticsProperty<ProviderBase<Object, T>>(
+        'provider',
+        (_providerListenable as ProviderSelector<dynamic, T>).provider
```

Comment:
The cast won't do. We probably want a `if ( is ProviderSelector ) else `

Reply:
What I did is this : 
```dart
if (_providerListenable is ProviderSelector<dynamic, T>) {
      properties.add(DiagnosticsProperty<ProviderBase<Object, T>>(
          'provider',
          (_providerListenable as ProviderSelector<dynamic, T>).provider
              as ProviderBase<Object, T>));
    } else {
      properties.add(DiagnosticsProperty<ProviderBase<Object, T>>(
          'provider', _providerListenable as ProviderBase<Object, T>));
    }
```

Not sure if it's the right one